File: week3/p7.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def removeCycle(head):
    slow, fast = head, head
    cycleFound = False

    # Detect cycle
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow == fast:
            cycleFound = True
            break

    if not cycleFound:
        return True  # No cycle present

    # Find start of cycle
    slow = head
    prev = None
    while slow != fast:
        prev = fast
        slow = slow.next
        fast = fast.next

    # Break the cycle
    prev.next = None
    return True

# Example: 1 -> 3 -> 4, pos = 2
head = Node(1)
head.next = Node(3)
head.next.next = Node(4)
head.next.next.next = head.next  # loop

print(str(removeCycle(head)).lower())

File: week3/p7.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

struct Node {
    int data;
    struct Node* next;
};

struct Node* newNode(int data) {
    struct Node* node = (struct Node*)malloc(sizeof(struct Node));
    node->data = data;
    node->next = NULL;
    return node;
}

bool removeCycle(struct Node* head) {
    struct Node *slow = head, *fast = head;
    bool cycleFound = false;

    // Detect cycle
    while (fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;
        if (slow == fast) {
            cycleFound = true;
            break;
        }
    }

    if (!cycleFound) return true; // No cycle present

    // Find start of cycle
    slow = head;
    struct Node* prev = NULL;
    while (slow != fast) {
        prev = fast;
        slow = slow->next;
        fast = fast->next;
    }

    // Break the cycle
    prev->next = NULL;
    return true;
}

int main() {
    struct Node* head = newNode(1);
    head->next = newNode(3);
    head->next->next = newNode(4);
    head->next->next->next = head->next; // loop to 2nd node

    printf("%s\n", removeCycle(head) ? "true" : "false");
    return 0;
}
